qemu/tests/throttle_block_set_io_throttle.py

```diff
@@ -0,0 +1,88 @@
+"""block_set_io_throttle testing on iothread enabled disk"""
+import time
+import json
+
+from virttest import utils_misc
+from virttest import utils_disk
+from virttest.utils_misc import get_linux_drive_path
+from virttest import error_context
+
+
+# This decorator makes the test function aware of context strings
+@error_context.context_aware
+def run(test, params, env):
+    """
+    block_set_io_throttle testing on iothread enabled disk
+
+    1) Boot vm with iothread enabled data disks.
+    2) Login guest and start io on data disk.
+    3) Execute block_set_io_throttle on data disks with different value.
+    4) Verify vm still alive.
+    """
+
+    def _get_window_disk_index_by_serail(serial):
+        cmd = "wmic diskdrive where SerialNumber='%s' get Index,Name"
+        disks = session.cmd_output(cmd % serial)
+        info = disks.splitlines()
+        if len(info) > 1:
+            attr = info[1].split()
+            return attr[0]
+        test.fail("Not find expected disk ")
+
+    def _execute_set_io_throttle(monitor):
+        cmd_qmp = params["cmd_qmp"]
+        throttle_value = params["throttle_value"].split(",")
+        images = params["data_images"].split()
+        for repeat in range(params.get_numeric("repeat_times", 2)):
+            for value in throttle_value:
+                logger.info("Start %s block_set_io_throttle %s", repeat, value)
+                for img in images:
+                    dev = img
+                    if params["drive_format_%s" % img] == "virtio":
+                        dev = "/machine/peripheral/%s/virtio-backend" % img
+                    cmd = cmd_qmp % (dev, value)
+                    logger.info(cmd)
+                    monitor.cmd_qmp("block_set_io_throttle", json.loads(cmd))
+                time.sleep(3)
+
+    logger = test.log
+    os_type = params["os_type"]
+    guest_cmd = params["guest_cmd"]
+    disk_serial = params["image_stg1"]
+    timeout = params.get_numeric("login_timeout", 360)
+    error_context.context("Get the main VM", logger.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+    qmp_monitor = vm.get_monitors_by_type("qmp")
+    if qmp_monitor:
+        qmp_monitor = qmp_monitor[0]
+    else:
+        test.error('Could not find a QMP monitor, aborting test')
+
+    logger.info("Execute io in guest...")
+    if os_type == 'windows':
+        img_size = params.get("image_size_stg1")
+        guest_cmd = utils_misc.set_winutils_letter(session, guest_cmd)
+        disk = _get_window_disk_index_by_serail(disk_serial)
+        utils_disk.update_windows_disk_attributes(session, disk)
+        test.log.info("Formatting disk:%s", disk)
+        driver = utils_disk.configure_empty_disk(session, disk, img_size,
+                                                 os_type)[0]
+        output_path = driver + ":\\test.dat"
```

Comment:
I am not sure understand what you mean, at line 70 the driver should a valid value or exception raised.
so i think it looks like redundant "" if not driver else driver + ":\\test.dat". what do you think ?

Reply:
Yes, I mean that I am afraid the return of `configure_empty_disk` would be null for windows guest.
```
avocado-vt/virttest/utils_disk.py
def configure_empty_disk(session, did, size, ostype, start="0M", n_partitions=1,
                         fstype=None, labeltype=PARTITION_TABLE_TYPE_MBR,
                         timeout=360):
        ...
        if ostype == "windows":
        return configure_empty_windows_disk(session, did, size, start,
                                            n_partitions, fstype,
                                            labeltype, timeout)
        ...

def configure_empty_windows_disk(session, did, size, start="0M",
                                 n_partitions=1, fstype="ntfs",
                                 labeltype=PARTITION_TABLE_TYPE_MBR,
                                 timeout=360, quick_format=True):
        ...
        drive_letter = set_drive_letter(session, did, partition_no=i + 1)
        if not drive_letter:
            return []
       ...
```
Might the probablitiy of this situation is very low. So your codes also look good to me.

Reply:
yes, [][0] will raise exception if configure_empty_windows_disk failed.

Reply:
> yes, [][0] will raise exception if configure_empty_windows_disk failed.

yes, I mean that the `set_drive_letter` could return `None` then the `configure_empty_windows_disk` will return `[]`
so that the `configure_empty_disk` returns the `[]`
At last `output_path = driver + ":\\test.dat"` will raise error, like `TypeError: can only concatenate list (not "str") to list`
Anyway, it will raise error before the next step.
LGTM.